File: app/src/main/java/com/drdisagree/colorblendr/ColorBlendr.java
package com.drdisagree.colorblendr;

import android.app.Application;
import android.content.Context;

import com.drdisagree.colorblendr.provider.RootServiceProvider;
import com.drdisagree.colorblendr.service.IRootService;
import com.google.android.material.color.DynamicColors;

import java.lang.ref.WeakReference;

public class ColorBlendr extends Application {

    private static ColorBlendr instance;
    private static WeakReference<Context> contextReference;

    public static Context getAppContext() {
        if (contextReference == null || contextReference.get() == null) {
            contextReference = new WeakReference<>(ColorBlendr.getInstance().getApplicationContext());
        }
        return contextReference.get();
    }

    private static ColorBlendr getInstance() {
        if (instance == null) {
            instance = new ColorBlendr();
        }
        return instance;
    }

    public static IRootService getRootService() {
        return RootServiceProvider.getRootServiceProvider();
    }

    public void onCreate() {
        super.onCreate();
        instance = this;
        contextReference = new WeakReference<>(getApplicationContext());
        DynamicColors.applyToActivitiesIfAvailable(this);
    }
}

File: app/src/main/java/com/drdisagree/colorblendr/ColorBlendr.kt
package com.drdisagree.colorblendr

import android.app.Application
import android.content.Context
import com.drdisagree.colorblendr.provider.RootConnectionProvider
import com.drdisagree.colorblendr.provider.ShizukuConnectionProvider
import com.drdisagree.colorblendr.service.IRootConnection
import com.drdisagree.colorblendr.service.IShizukuConnection
import com.drdisagree.colorblendr.utils.app.CrashLogger
import com.google.android.material.color.DynamicColors
import io.github.muntashirakon.adb.PRNGFixes
import org.lsposed.hiddenapibypass.HiddenApiBypass
import java.lang.ref.WeakReference

class ColorBlendr : Application() {

    override fun onCreate() {
        super.onCreate()
        instance = this
        contextReference = WeakReference(applicationContext)
        DynamicColors.applyToActivitiesIfAvailable(this)
        PRNGFixes.apply()
        CrashLogger.install(this)
    }

    override fun attachBaseContext(base: Context) {
        super.attachBaseContext(base)
        HiddenApiBypass.addHiddenApiExemptions("L")
    }

    companion object {
        private lateinit var instance: ColorBlendr
        private lateinit var contextReference: WeakReference<Context>

        val appContext: Context
            get() {
                if (!this::contextReference.isInitialized || contextReference.get() == null) {
                    contextReference = WeakReference(
                        getInstance().applicationContext
                    )
                }
                return contextReference.get()!!
            }

        fun initializeForPreview(context: Context) {
            if (!this::contextReference.isInitialized || contextReference.get() == null) {
                contextReference = WeakReference(context.applicationContext ?: context)
            }
        }

        private fun getInstance(): ColorBlendr {
            if (!this::instance.isInitialized) {
                instance = ColorBlendr()
            }
            return instance
        }

        val rootConnection: IRootConnection?
            get() = RootConnectionProvider.getServiceProvider

        val shizukuConnection: IShizukuConnection?
            get() = ShizukuConnectionProvider.getServiceProvider
    }
}